README.md
RASCAL
============================================================

### Robotic Autonomous Space Cadet Administration Lackey

<p align="center">
  <img src="https://i.imgur.com/jmm8B6z.jpg" alt="I am RASCAL" height="250"/>
  </br>
  <sub><strong>Robotic Autonomous Space Cadet Administration Lackey</strong></sub>
</p>

---

RASCAL is a customized administration bot App for Facebook Groups, designed as a rearchitected update and expansion of [ZacSweers' Facebook Modbot](https://github.com/ZacSweers/FB_Mod_Bot). RASCAL's design is intended to aid in the speed, consistency, security, and reduction of repetitive human-led administration duties; it is **not** intended to be a replacement for administrative oversight or involvement in Facebook Groups.

### About RASCAL

RASCAL runs as a Facebook Developer App, is coded in Python 3.7 using the [facepy](https://github.com/jgorset/facepy) wrapper for Facebook Graph API, and has the following prerequisites:

- **Facebook Account for Bot:** you can set up an entirely different account for your RASCAL bot, or use your personal account if you don't require them to be separated.
- **Facebook Page for Bot:** you should create a page that will operate as the identity and administration portal and granting access tokens
- **Add Page as Group Administrator:** adding the Facebook page as an administrator for the group you want the bot to run in will allow the bot to handle assigned administrative actions
- **Register Account for Facebook Developer:** in order to sign up for Facebook Developer, you will have to go through the application process provided on the Facebook Developer website

*Why make a dedicated app instead of just creating an access token?*

Only an app (with an app ID and secret key) can programatically request an **extended** access token. Without doing this, you'd need to generate a new access token every two hours.

### RASCAL Modifications

You may modify RASCAL to perform administrative taskmastering in individual Facebook groups, so long as RASCAL's updates and endeavors abide Asimov's Three Laws of Robotics:

1. A robot may not injure a human being or, through inaction, allow a human being to come to harm.
2. A robot must obey orders given it by human beings except where such orders would conflict with the First Law.
3. A robot must protect its own existence as long as such protection does not conflict with the First or Second Law.

### RASCAL To-Do

* Upgrade from Python 2 -> Python 3.7
* Validate facepy API Changes and Updates
* Windows / Linux Integration

* Member Management / Welcome
* Post Approval
* Administrator Protection
* Spam Post Detection / Clean-Up
* Fake Account Detection / Clean-Up
* Descriptive Comments on Images
* Interactive Replies to Comment Tags
* Database of Curated Collections Shared by Members

* Documentation

### Known Issues

* Outdated Codebase


### [Changelog](https://github.com/WhitneyOnTheWeb/RASCAL/blob/master/CHANGELOG.md)

CHANGELOG.md
RASCAL Changelog
================

### 02/08/14 (ZS)

- Created changelog
- Added `[PARKING]` tag checking

### 26/06/2019 (WK)

- 26/06/2019 (v1.0):
    - Forked repro of old bot used for automodding roommate search FB groups
    - Update README and directory structure
    - Rearchitect from Python 2 -> 3.7
      - Updated runtime and requirements files
      - Add scaffolding for Windows/Mac features
    - Redesign for generalized group administration features
    - Design customized TOCA group administrative features
    - Move property and cache related functions into util.py

rascal_bot/util.py
import sys
import csv
import cPickle as pickle
import subprocess

'''----------------------------------------------------------------------------
Project:    RASCAL: Robotic Autonomous Space Cadet Adminstration Lackey
File:       util.py
Author:     @WhitneyOnTheWeb

Original:   ZacSweers FB Modbot (2014 / Py2)

General utility functions for RASCAL operation
----------------------------------------------------------------------------'''
__author__ = 'Whitney King'


'''-------------------------------------
Global Variables
================
'''
time_limit = 86400                  # 60 * 60 * 24 (s --> h)
prop_file = 'properties'            # properties pickle name
warned_db = 'fb_subs_cache'         # warned posts cache
valid_db = 'fb_subs_valid_cache'    # valid posts cache

tag_delimiters = ('*', '-')         # leading symbols for post tags
post_tags = load_list('post_tags.txt')

bot_delimiters = ('!', '#')         # leading symbols for bot tags
bot_tags = load_list('bot_tags.txt')

# Booleans
is_heroku = False                    # is bot running on heroku?
is_pickled = False                   # is there a .pkl for properties?
dry_run = False                     # disable deletion dry-run
extend_key = False                  # is extended key


'''-------------------------------------
Class: Color

Text color formatting for logging output
'''
class Color:
    PURPLE = '\033[95m'
    CYAN = '\033[96m'
    DARKCYAN = '\033[36m'
    BLUE = '\033[94m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    END = '\033[0m'


'''-------------------------------------
Method: test

Template placeholder function for displaying
output when running tests
'''
def test():
    log('Test', Color.PURPLE)


'''-------------------------------------
Method: extend_access_token

Used to automatically extend access token
'''
def extend_access_token(graph, now_time, saved_props,
                        sublets_api_id, sublets_secret_key):
    log("Extending access token", Color.BOLD)
    result = graph.extend_access_token(sublets_api_id, sublets_secret_key)
    new_token = result['access_token']
    new_time = int(result['expires']) + now_time
    saved_props['sublets_oauth_access_token'] = new_token
    saved_props['access_token_expiration'] = new_time
    log("Token extended", Color.BOLD)


'''-------------------------------------
Method: load_list

Generate list of values to be used by RASCAL
- Reads from a .txt file into a list variable

Input:
    <lfile>: name of file containing list with
             syntax (one per line):
             value
Return:
    <list>: with values from file
'''
def load_list(lfile):
    vals = []
    with open('./files/{}'.format(lfile), 'r') as f:
        lst = csv.reader(f)     # get text from file
        for val in lst:         # read each value
            vals.append(val)    # add value to list
    f.close()                   # close file when done
    return vals


'''-------------------------------------
Method: get_settings

Generate settings dictionary with key-value
pairs stored as items in a .txt file

Input:
    <pfile>: name of file with settings using
             syntax (one per line):
             key: value
Return:
    <dict>:  dictionary with settings
             key-value pairs
'''
def get_settings(pfile):
    prop = {}
    with open('./files/{}'.format(pfile), 'r') as keys:
        for key in keys:
            k, v = key.split(': ')
            if v.endswith('\n'): v = v[:-1]
            prop[k] = v
    keys.close()
    return prop


'''-------------------------------------
Method: init_properties

Initializes property values for RASCAL,
then uploads them in chosen format
'''

def init_properties():
    test_dict = get_settings('properties.txt')
    save_properties(test_dict)

    saved_dict = load_properties()
    assert test_dict == saved_dict


'''-------------------------------------
Method: save_properties

Save RASCAL properties to either:
 - memcache (heroku)
 - pickle (local file)
 - text file

Input:
    <data>: dict of property key-value pairs
'''
def save_properties(data):
    if is_heroku:        # save to heroku
        mc.set('properties', data)
    else:               # save to .pkl
        with open(prop_file, 'w+') as f:
            pickle.dump(data, f)
        f.close()
                        # save to .txt
    # with open('./files/properties.txt', 'w') as f:


'''-------------------------------------
Method: load_properties

Loads saved property values from either:
 - memcache (heroku)
 - pickle (local file)
 - text file

 Return:
    <dict>:  dictionary with properties
             key-value pairs
'''
def load_properties():
    if is_heroku:        # load from heroku
        obj = mc.get('properties')
        if not obj:
            return {}
        else:
            return obj
    else:               # load from .pkl
        if os.path.isfile(prop_file):
            with open(prop_file, 'r+') as f:
                data = pickle.load(f)
                return data
        else:           # load from .txt
            return get_settings('properties.txt')


'''-------------------------------------
Method: set_property

Sets key in property dictionary to a
user defined value, updates value in
properties text file

Input:
    <props>:    dict containing properties
    <key>:      key to add or update
    <value>:    value to set
'''
def set_property(props, key, value):
    # Add if needed for RASCAL operation
    pass


'''-------------------------------------
Method: load_cache

Loads cache from user defined locations.
Either returns cached values or original data

Input:
    <cachename>:    cache to select
    <data>:         data to return

Returns:
    <obj>:          object with cached values
'''
def load_cache(cachename, data):
    if is_heroku:
        if is_heroku:
            obj = mc.get(cachename)
            if not obj:
                return data
            else:
                return obj
    else:
        if os.path.isfile(cachename):
            with open(cachename, 'r+') as f:
                # If the file isn't at its end or empty
                if f.tell() != os.fstat(f.fileno()).st_size:
                    return pickle.load(f)
        else:
            log("--No cache file found. Creating new cache file.", Color.BLUE)
            return data


'''-------------------------------------
Method: save_cache

Saves cache to user defined location.

Input:
    <cachename>:    cache to select
    <data>:         data to save
'''
def save_cache(cachename, data):
    if is_heroku:
        mc.set(cachename, data)
    else:
        with open(cachename, 'w+') as f:
            pickle.dump(data, f)


'''-------------------------------------
Method: log

Checks for Color class, then prints log output

Input:
    <message>: logging message to display
'''
def log(message, *colorargs):
    if len(colorargs) > 0:
        print(colorargs[0] + message + Color.END)
    else:
        print (message)



'''-------------------------------------
Method: notify

Currently sends notifications on Mac

- Needs updates for Windows/Linux
'''
def notify():
    if sys.platform == "darwin":
        try:
            subprocess.call(
                ["terminal-notifier", "-message", "Tests done", "-title",
                 "FB_Bot", "-sound", "default"])
        except OSError:
            print('If you have terminal-notifier, this would notify')


'''-------------------------------------
Method: read_lines

Reads in multi-line strings based on "newline" type
http://stackoverflow.com/a/16260159/3034339

- Needs to be validated
'''
def read_lines(f, newline):
    buf = ""
    while True:
        while newline in buf:
            pos = buf.index(newline)
            yield buf[:pos]
            buf = buf[pos + len(newline):]
        chunk = f.read(4096)
        if not chunk:
            yield buf
            break
        buf += chunk

rascal_bot/TestPosts.py
__author__ = 'wking'

'''Add new test posts here '''


# Replace these examples posts from hzsweers with samples relevant to Facebook group

good_posts = [
    """[OFFERING]

Charming 1-bedroom studio apartment in Hyde Park/North Campus on UT Shuttle stop!

<hidden>

*Quiet, downstairs corner unit next to private courtyard
*UT Shuttle (IF) and Cap Metro #5 stop 20 seconds from your door
*Small, 10-unit complex, renovated in 2013
*Walk-in closet
*Wood floors
*Nonfurnished
*Nonsmoking
*Central AC/heat
*On-site laundry
*Official sublease through the property manager
*Parking space close to unit
*Trash, water paid
*Walk to CVS, The Triangle, Daily Juice, Avenue B Grocery, tennis courts

$770/mo for 12 month lease
Available for Mid-August move-in

*$75 application fee and $300 deposit*
    """,

    """[OFFERING]
-Fall 14, Spring 15
-For a shared bedroom and private bath at The Block on 221/2 and Pearl.  (2 by 2 apartment)
-535 per month or 636 with parking.
-Message me ASAP if interested
-It is for the room on the left. The other girl sharing that room is SUUPER cool!
-water and electricity not included
-internet/cable included
-W/D in unit
-large balcony
-bus stop super close
    """,

    """[OFFERING]

Passing my sub-lease in The Block on 28th, apartment <hidden>.
Due to a family emergency, I will not be attending UT, and I need someone to take over my lease as soon as possible.
It is two bedroom, 2 restroom apartment in west campus, and walking distance from UT (8-15 min). The rent is $445/month plus utilities, and parking space is additional for an extra of $100/month. You will be sharing one bedroom with a girl, and there are two other girls also living in the apartment.
The apartment already includes basic cable, wifi buildings, and internet.
    """,

    """[OFFERING]
-FIRST MONTH'S RENT IS FREE ($899 per month becomes $824 per month!)
-Subleasing my room at the Villas on Guadalupe for the Fall 2014 - Spring 2015 school year.
-You get your own room, private bath, and parking spot (optional)
-It is a 2 room apartment w/ loft, w/ 3 other roommates (males)
-Rent is $899 per month, parking is available and is an optional additional $65.
-Washer/dryer included in the unit
-Unfurnished
-Roommates don't smoke
-Right above Torchy's Tacos and across the street from Wataburger
-Apartment complex has a pool and weight room.
- Move in date is Aug 16th.
- Bedroom A in the layout photo.
Message if interested or for more details. Thanks!
    """,

    """[LOOKING] [ROOMING]

-20-year-old female, government major- Italian, exchange student from France.
-Looking for a studio/ room + private bathroom if possible
-For the full 2014-2015 academic year
-Close to campus
-Available right now if possible / max from the 17th of August
-Around $750/ $800
-Furnished if possible.
Thanks :)
    """,

    """[OFFERING] 2BR/1BA in West Campus $865/mo. for ENTIRE UNIT (that's only $432.50 per room). Lease term is August 2014-2015. It isn't furnished, but I can leave a desk, some shelves, etc. for free if you'd like. This lease is for the entire unit, so you must have/find your own roommate. Utilities are not included, but it has free basic cable (with only one outlet, sadly). This will be an official lease sign-over through the complex. This is the cheapest place to live in West Campus with your own, private room. That's why I've lived there for the last 2 years. Location and price are amazing.
Here is the Craigslist ad with lots of details and pictures: <hidden>
    """
]

bad_posts = [
    """[OFFERING] (UniversityEstates)
-Looking to sublease my apartment
-located <hidden>
- rent is $500 a month including ALL utilities.
-Renter's insurance is required, which is $10-25 a month depending where you get it
-Lease starts August 22, 2014 - July 31st, 2015
-unfurnished
-4x4
-co-ed
-Close to downtown
-UT shuttle passes the front of the complex often
-Includes parking
- Each room has it's own lease
-Official sublease
-Full kitchen and laundry
-Apartment complex contains pool, gym, computer lab, clubhouse.
- http://www.universityestatesataustin.com/index.php
For more info go ahead an message me through here! I'll be checking my other folder daily, and i'll be more than glad to help!
    """,

    """[OFFERING] 2BR/1BA in West Campus $865/mo. for ENTIRE UNIT (randomwordhere) (that's only $432.50 per room). Lease term is August 2014-2015. It isn't furnished, but I can leave a desk, some shelves, etc. for free if you'd like. This lease is for the entire unit, so you must have/find your own roommate. Utilities are not included, but it has free basic cable (with only one outlet, sadly). This will be an official lease sign-over through the complex. This is the cheapest place to live in West Campus with your own, private room. That's why I've lived there for the last 2 years. Location and price are amazing.
Here is the Craigslist ad with lots of details and pictures: <hidden>
    """,

    """[OFFERING][ROOMING] 2BR/1BA in West Campus $865/mo. for ENTIRE UNIT (randomwordhere) (that's only $432.50 per room). Lease term is August 2014-2015. It isn't furnished, but I can leave a desk, some shelves, etc. for free if you'd like. This lease is for the entire unit, so you must have/find your own roommate. Utilities are not included, but it has free basic cable (with only one outlet, sadly). This will be an official lease sign-over through the complex. This is the cheapest place to live in West Campus with your own, private room. That's why I've lived there for the last 2 years. Location and price are amazing.
Here is the Craigslist ad with lots of details and pictures: <hidden>
    """
]

rascal_bot/post_cleanup.py
#!/usr/bin/env python
# coding=utf-8
import csv
import datetime
import getopt
import json
import os
import re
import sys
import time

import facepy

import util as u

'''----------------------------------------------------------------------------
Project:    RASCAL: Robotic Autonomous Space Cadet Adminstration Lackey
File:       post_cleanup.py
Author:     @WhitneyOnTheWeb

Original:   check_and_delete.py

Functionality designed to review posts for spam, fraud, or other malicious
patterns, then flag / warn / action identified violations

- Notes:  it seems that a good number of these methods can be moved
          over into util.py to be used in a more generalized manner
          with RASCAL functionality
----------------------------------------------------------------------------'''
__author__ = 'Whitney King'


'''-------------------------------------
Global Variables
================
'''
time_limit = 86400                  # 60 * 60 * 24 (s --> h)
warned_db = 'fb_subs_cache'         # warned posts cache
valid_db = 'fb_subs_valid_cache'    # valid posts cache

tag_delimiters = ('*', '-')         # leading symbols for post tags
post_tags = u.load_list('post_tags.txt')

bot_delimiters = ('!', '#')         # leading symbols for bot tags
bot_tags = u.load_list('bot_tags.txt')

# Booleans
dry_run = False                     # disable deletion dry-run
extend_key = False                  # is extended key


# Manually update API token
def update_token(token):
    log("Updating token", Color.BLUE)
    graph = facepy.GraphAPI(token)
    try:
        graph.get('me/posts')
        props_dict = load_properties()
        props_dict['sublets_oauth_access_token'] = token
        props_dict['access_token_expiration'] = time.time() + 7200  # 2 hours buffer
        save_properties(props_dict)
        log("Token updated, you should now extend it", Color.BLUE)
    except Exception as e:
        log("API error - " + e.message, Color.RED)


# Set a property by name
def update_prop(prop_name, value):
    if prop_name in ("sublets_oauth_access_token", "access_token_expiration"):
        log("Please use -u or -e to update or extend tokens", Color.RED)
        return
    log("Setting \"" + prop_name + "\" to \"" + value + "\"", Color.BLUE)
    props_dict = load_properties()
    if prop_name not in props_dict.keys():
        input_response = raw_input("This key doesn't exist, do you want to add it? Y/N")
        if input_response.lower() != "y":
            return
    props_dict[prop_name] = value
    save_properties(props_dict)
    log("Done setting \"" + prop_name + "\"", Color.BLUE)


# Method for checking tag validity
def get_tags_old(message_text):
    p = re.compile(
        "^(-|\*| )*([\(\[\{])((looking)|(rooming)|(offering)|(parking))([\)\]\}])(:)?(\s|$)",
        re.IGNORECASE)

    split = message_text.strip().split(" ")

    # Check that they're on the first line
    if not p.match(split[0]) and split[0] not in allowed_leading_characters:
        return None

    tags_list = [l.lower()[1:-1] for l in split for m in (p.search(l),) if m]
    if len(tags_list) > 0:
        return tags_list
    else:
        return None


# Method for checking tag validity
def get_tags(message_text):
    p = re.compile(
        r"^(-|\*| )*(([\(\[\{])(.+)([\)\]\}]))+(:)?(\s|$)",
        re.IGNORECASE)

    # Insert space between two consecutive tags
    message_text = re.sub(r"([\)\]\}])([\(\[\{])", r"\1 \2", message_text)
    firstline_split = message_text.split("\n")[0].strip().split()

    # Check that they're on the first line
    if not p.match(firstline_split[0]) and firstline_split[0] not in allowed_leading_characters:
        return None

    tags_list = [re.sub(r"^(-|\*| )*", "", full_tag.lower())[1:(-2 if full_tag[-1] == ":" else -1)] for full_tag in
                 firstline_split for matched in (p.search(full_tag),) if matched]

    # Should really learn how to do python's builtin logging...
    # log('--Tags: ' + ', '.join(tags_list), Color.BLUE)

    if len(tags_list) > 0 and set(tags_list).issubset(allowed_tags):
        return tags_list
    else:
        return None


# Can't have "rooming" AND "offering". These people are usually just misusing the rooming tag
def validate_tags(tags):
    if not tags:
        return False
    elif "rooming" in tags and "offering" in tags:
        return False
    else:
        return True


# Method for checking if pricing reference is there
def check_price_validity(message_text):
    p = re.compile(
        "(\$)|((\d)+( )?((per)|(/)|(a))( )?(/)?((month)|(mon)|(mo))(\s)?)",
        re.IGNORECASE)

    if re.search(p, message_text) is not None:
        return True
    else:
        return False


# Checking if there's a parking tag
def check_for_parking_tag(message_text):
    p = re.compile(
        "^(-|\*| )*([\(\[\{])(parking)([\)\]\}])(:)?(\s|$)",
        re.IGNORECASE)

    if re.search(p, message_text):
        return True
    else:
        return False


# Method for extending access token
def extend_access_token(saved_props, token, sublets_api_id,
                        sublets_secret_key):
    log("Extending access token", Color.BOLD)
    access_token, expires_at = facepy.get_extended_access_token(
        token,
        sublets_api_id,
        sublets_secret_key
    )
    new_token = access_token
    unixtime = time.mktime(expires_at.timetuple())
    print time.mktime(expires_at.timetuple())
    saved_props['sublets_oauth_access_token'] = new_token
    saved_props['access_token_expiration'] = unixtime
    log("Token extended", Color.BOLD)


# Method for retrieving user ID's of admins in group, ignoring bot ID
def retrieve_admin_ids(group_id, auth_token):
    # Retrieve the uids via FQL query
    graph = facepy.GraphAPI(auth_token)
    admins_query = \
        "SELECT uid FROM group_member WHERE gid=" + group_id + " AND" + \
        " administrator"
    admins = graph.fql(query=admins_query)

    # Parse out the uids from the response
    admins_list = [admin['uid'] for admin in admins]

    # Update the admin_ids in our properties
    saved_props = load_properties()
    saved_props['admin_ids'] = admins_list
    save_properties(saved_props)

    return admins_list


# Delete posts older than 30 days old
def delete_old_posts(graph, group_id, admin_ids):
    old_date = int(time.time()) - 2592000  # 30 days in seconds
    old_query = "SELECT post_id, message, actor_id FROM stream WHERE " + \
                "source_id=" + group_id + " AND created_time<" + str(old_date) + \
                " LIMIT 300"
    log("Getting posts older than:")
    log("\t" + datetime.datetime.fromtimestamp(old_date)
        .strftime('%Y-%m-%d %H:%M:%S'))
    posts = graph.fql(query=old_query)
    deleted_posts_count = 0
    for post in posts["data"]:
        post_id = post['post_id']
        actor_id = post['actor_id']
        if int(actor_id) in admin_ids:
            # log('\n--Ignored post: ' + post_id, Color.BLUE)
            continue
        print post_id
        graph.delete(post_id)
        deleted_posts_count += 1
    log("Deleted " + str(deleted_posts_count) + " old posts", Color.RED)


# Main runner method
def sub_group():
    # Load the properties
    saved_props = load_properties()

    # Access token
    sublets_oauth_access_token = saved_props['sublets_oauth_access_token']

    # Access token expiration
    access_token_expiration = saved_props['access_token_expiration']

    # API App ID
    sublets_api_id = saved_props['sublets_api_id']

    # API App secret key
    sublets_secret_key = saved_props['sublets_secret_key']

    # ID of the FB group
    group_id = saved_props['group_id']

    # IDs of admins (unused right now, might remove later)
    admin_ids = saved_props['admin_ids']

    # FQL query for the group
    group_query = "SELECT post_id, message, actor_id FROM stream WHERE " + \
                  "source_id=" + group_id + " LIMIT 50"

    # Get current time
    now_time = time.time()

    # For logging purposes
    log("CURRENT CST TIMESTAMP: " + datetime.datetime.fromtimestamp(
        now_time - 21600).strftime('%Y-%m-%d %H:%M:%S'), Color.UNDERLINE)

    # Make sure the access token is still valid
    if access_token_expiration < now_time:
        sys.exit("API Token is expired")

    # Warn if the token's expiring soon
    if access_token_expiration - now_time < 604800:
        log("Warning - access token expires in less than a week", Color.RED)
        log("-- Expires on " + datetime.datetime.fromtimestamp(
            access_token_expiration).strftime('%Y-%m-%d %H:%M:%S'))

        # If you want it to automatically when it's close to exp.
        global extend_key
        extend_key = True

    # Extend the access token, default is ~2 months from current date
    if extend_key:
        extend_access_token(saved_props, sublets_oauth_access_token, sublets_api_id,
                            sublets_secret_key)

    # Log in, try to get posts
    graph = facepy.GraphAPI(sublets_oauth_access_token)

    # Make our first request, get the group posts
    group_posts = graph.fql(query=group_query)

    # Load the pickled cache of valid posts
    valid_posts = []
    log("Checking valid cache.", Color.BOLD)
    valid_posts = load_cache(valid_db, valid_posts)
    log('--Valid cache size: ' + str(len(valid_posts)), Color.BOLD)
    invalid_count = 0

    # Loop over retrieved posts
    for post in group_posts["data"]:

        # Important data received
        post_message = post['message']  # Content of the post
        post_id = post['post_id']  # Unique ID of the post

        # Unique ID of the person that posted it
        actor_id = post['actor_id']

        # Ignore mods and certain posts
        if int(actor_id) in admin_ids:
            log('\n--Ignored post: ' + post_id, Color.BLUE)
            continue

        # Boolean for tracking if the post is valid
        valid_post = True

        # Log the message details
        # log("\n" + post_message[0:75].replace('\n', "") + "...\n--POST ID: " +
        # str(post_id) + "\n--ACTOR ID: " + str(actor_id))

        # Check for pricing
        if not check_price_validity(post_message):
            valid_post = False
            log('----$', Color.RED)
            invalid_count += 1

        # Check for tag validity, including tags that say rooming and offering
        tags = get_tags(post_message)
        if not validate_tags(tags):
            valid_post = False
            log('----Tag', Color.RED)
            invalid_count += 1

        # Check post length.
        # Allow short ones if there's a craigslist link or parking
        if len(post_message) < 200 \
                and "craigslist" not in post_message.lower() \
                and not check_for_parking_tag(post_message):
            valid_post = False
            log('----Length', Color.RED)
            invalid_count += 1

        # Not a valid post
        if not valid_post:
            if dry_run:
                log("Dry - invalid deletion", Color.RED)
                log("--ID: " + post_id, Color.RED)
                log("--Message: " + post_message, Color.RED)
                log("\n")
            else:
                graph.delete(post_id)
        else:
            valid_posts.append(post_id)

    if not dry_run:
        log("Deleted " + str(invalid_count) + " invalid posts", Color.RED)

    # # Delete posts older than 30 days
    delete_old_posts(graph, group_id, admin_ids)

    # Save the updated caches
    log('Saving valid cache', Color.BOLD)
    save_cache(valid_db, valid_posts)

    save_properties(saved_props)

    # Done
    notify_mac()


# Main method
if __name__ == "__main__":

    try:
        opts, args = getopt.getopt(sys.argv[1:], "fdpesu:n:v:g:",
                                   ["flushvalid", "dry", "printprops", "extend", "setprops", "token=", "propname=",
                                    "propvalue=", "propname="])
    except getopt.GetoptError:
        print 'check_and_delete.py -f -d -p -e -s -u <token> -n <propname> -v <propvalue> -g <propname>'
        sys.exit(2)

    # Check to see if we're running on Heroku
    if os.environ.get('MEMCACHEDCLOUD_SERVERS', None):
        import bmemcached

        log('Running on heroku, using memcached', Color.BOLD)

        # Authenticate Memcached
        running_on_heroku = True
        mc = bmemcached.Client(os.environ.get('MEMCACHEDCLOUD_SERVERS').split(','),
                               os.environ.get('MEMCACHEDCLOUD_USERNAME'), os.environ.get('MEMCACHEDCLOUD_PASSWORD'))

    propname = None
    propval = None
    log("Args - " + str(opts), Color.BOLD)
    if len(opts) != 0:
        for o, a in opts:
            if o in ("-e", "--extend"):
                extend_key = True
            elif o in ("-d", "--dry"):
                dry_run = True
            elif o in ("-s", "--setprops"):
                set_new_props()
                sys.exit()
            elif o in ("-u", "--update"):
                update_token(a)
                sys.exit()
            elif o in ("-n", "--propname"):
                propname = a
            elif o in ("-v", "--propvalue"):
                propval = a
            elif o in ("-p", "--printprops"):
                log("Printing props", Color.BLUE)
                props = load_properties()
                print props.keys()
                sys.exit()
            elif o in ("-g", "--getprop"):
                log("Getting value for " + a, Color.BLUE)
                props = load_properties()
                if a not in props.keys():
                    sys.exit(a + " doesn't exist in props")
                print props[a]
                sys.exit()
            elif o in ("-f", "--flushvalid"):
                response = raw_input("Are you sure? Y/N")
                if response.lower() == 'y':
                    log("Flushing cache for valid", Color.BLUE)
                    save_cache(valid_db, [])
                    log("Flushed cache", Color.BLUE)
                sys.exit()
            else:
                sys.exit('No valid args specified')

    if propname or propval:
        if propname and propval:
            log(propname + propval)
            update_prop(propname, propval)
        else:
            sys.exit('Must specify a prop name and value')
    else:
        sub_group()
